planning.py

```diff
@@ -294,6 +329,223 @@ def add_vertices(
                     if self.workspace.edge_is_in_collision(vo.coord, v.coord):
                         continue
                     vo.connect(v)
+        self.preprocessing_time = time.time() - start_time
+
+    def closest_point_not_in_collision(self, q, q_nearest):
+        """return closest point that is not in collision with workspace"""
+        while self.workspace.point_is_in_collision(q):
+            q = q + (q - q_nearest) / np.linalg.norm(q - q_nearest)
+        return q
+
+
+class RRT(RRG):
+    def __init__(self, workspace, q0):
+        super().__init__(workspace, q0)
+        self.v_start.parent = None
+
+    def query(
+        self,
+        start,
+        goal,
+        n=1000,
+        min_edge_len=0.5,
+        max_edge_len=1,
+    ):
+        new_size = self.graph.n + n
+        start_time = time.time()
+        while self.graph.n < new_size:
+            # TODO maybe change min and max edge len based on how many points have been taken?
+            # add radius with minimum path
+            self.expand_graph(min_edge_len, max_edge_len)
+            if self.end_condition(goal):
+                self.preprocessing_time = 0
+                self.query_time = time.time() - start_time
+                break
+        self.preprocessing_time = 0
+        self.query_time = time.time() - start_time
+
+    def expand_graph(self, min_edge_len, max_edge_len):
+        """Add a vertex to the RRT graph"""
+        q = self.workspace.sample()
+
+        v_nearest, dist = self.closest_vertex(q)
+        if dist < min_edge_len:
+            return
+        q_nearest = v_nearest.coord
+
+        # move toward q as much as possible
+        if dist > max_edge_len:
+            q = q_nearest + (q - q_nearest) * max_edge_len / dist
+        if self.workspace.point_is_in_collision(q):
+            q = self.closest_point_not_in_collision(q, q_nearest)
+        # don't add if edge is in collision
+        if self.workspace.edge_is_in_collision(q_nearest, q):
+            return
+
+        v = self.graph.add_vertex(q)
+        v.connect(v_nearest)
+        v.parent = v_nearest
+
+    def end_condition(self, goal, goal_dist=0.5):
+        v_nearest, dist = self.closest_vertex(goal)
+        if dist <= goal_dist and not self.workspace.edge_is_in_collision(
+            v_nearest.coord, goal
+        ):
+            v = self.graph.add_vertex(goal)
+            v.connect(v_nearest)
+            v.parent = v_nearest
+            self.v_goal = v
+            return True
+        return False
+
+    def draw(self, ax, vertices=True, edges= True, rgb=(1,0.5,0)):
+        super().draw(ax, vertices=vertices, edges=edges, rgb=rgb)
+        self.find_path(ax)
+    
+    def find_path(self, ax):
```

Comment:
I would prefer to keep drawing/visualization completion independent of querying the path. So the `find_path` method should have no drawing functionality, and the `draw` method should not call `find_path`. Instead, `draw` should just draw the tree itself without the path and `find_path` should return a list/array of coordinates for the nodes that make up the path, which be easily drawn separately. The reasoning for this is that often one would want to find the path without drawing it without.